### PR TITLE
Elliptic curve factorization

### PR DESCRIPTION
**Pull request by abh2k**
* ntheory
  * Implemented  Lenstra's Elliptic curve factorization

### CODE REVIEW

**Comment by abhinav28071999** [comment]
@jmig5776 @leosartaj @Yathartha22 This ecm should be working.

**Comment by jmig5776** [comment]
I will review it tonight.

File: sympy/ntheory/ecm.py
Comment (abhinav28071999) [inline sympy/ntheory/ecm.py:207]: I am doubtful about this. Increase bound should probably be removed.

File: sympy/ntheory/ecm.py
Comment (jmig5776) [inline sympy/ntheory/ecm.py:34]: Add in the documentation the mathematics of adding points in the Montogomery form. Also, use variable names more readable for eg. `s` -> `sum`. Also as I don't know much about mathematics of points in Montogomery form. I will try to check the logic after thoroughly understanding about it.

File: sympy/ntheory/ecm.py
Comment (jmig5776) [inline sympy/ntheory/ecm.py:54]: Same goes here.

File: sympy/ntheory/ecm.py
Comment (jmig5776) [reply sympy/ntheory/ecm.py:207]: Why?

File: sympy/ntheory/tests/test_ecm.py
Comment (jmig5776) [inline sympy/ntheory/tests/test_ecm.py:4]: Add more tests.

**Review by jmig5776 — COMMENTED** [review]
In most of the places the variable names used are random, change them such that the code and algorithm is more readable. Tests should be added more with commenting on the efficiency of this algorithm for eg (time difference). As far as the logic is concerned we have to check more deeply about the bound conditions. 
Also one more thing, please add description of PR and also add a link to the paper which you are adding in this PR.

File: sympy/ntheory/ecm.py
Comment (abhinav28071999) [reply sympy/ntheory/ecm.py:207]: I think `increase_bound` can be used but i am not sure about the multiplier used `*10` it increases too rapidly. It should be lowered but not sure to which number.

File: sympy/ntheory/tests/test_ecm.py
Comment (abhinav28071999) [reply sympy/ntheory/tests/test_ecm.py:4]: I have a ques. As we use random in ecm, smichr told me to use `seed` in the test cases. I am not sure how to use that.

Comment (abhinav28071999) [reply sympy/ntheory/tests/test_ecm.py:4]: Moreover, the main advantage of ecm is seen in bigger numbers >30digits and the timing of those test will be atleast 5~10secs and can be even more depending on the size.

File: sympy/ntheory/ecm.py
Comment (abhinav28071999) [reply sympy/ntheory/ecm.py]: The psedocode for Montgomery `addition`, `doubling`, `ladder` is from http://www.hyperelliptic.org/tanja/SHARCS/talks06/Gaj.pdf
And the `ecm`      function pseudocode is available in the book Prime numbers computational perspective. I have added these in references.

File: sympy/ntheory/tests/test_ecm.py
Comment (abhinav28071999) [reply sympy/ntheory/tests/test_ecm.py:4]: Currently i have added the test cases and it required around ~15-25secs to evaluate.

**Comment by abhinav28071999** [comment]
@jmig5776 i have added  the    references, please look into the pseudocodes.

**Comment by abhinav28071999** [comment]
@jmig5776 @Yathartha22 @leosartaj any updates?

**Comment by jmig5776** [comment]
@abhinav28071999 I will be busy for this week and the one next to it due to exams. I will review the pseudocode whenever I find it possible. And don't wait for the review, you can start working on the next thing.

File: sympy/ntheory/tests/test_ecm.py
Comment (jmig5776) [reply sympy/ntheory/tests/test_ecm.py:4]: Yes because the seed will create the same random numbers on multiple executions of the code. You can look at it from [here](https://www.geeksforgeeks.org/random-seed-in-python/).

If these tests take much time than you can add @slow tag to these tests.

File: sympy/ntheory/ecm.py
Comment (jksuom) [inline sympy/ntheory/ecm.py:18]: I would make this a method of `Point`. For example `def add(P, Q, diff,n):`. It is not necessary to change `P` to `self`.

File: sympy/ntheory/ecm.py
Comment (jksuom) [inline sympy/ntheory/ecm.py:44]: This could be `def double(P, n, a24)'.

File: sympy/ntheory/ecm.py
Comment (jksuom) [inline sympy/ntheory/ecm.py:69]: And here `P` should be the first parameter: `def mont_ladder(P, k, n, a24):`

File: sympy/ntheory/ecm.py
Comment (jksuom) [inline sympy/ntheory/ecm.py:165]: If `g == n`, then it looks like `Q.z` is zero mod `n` which means that `Q` is the neutral element (or zero) of the curve group. It cannot be used for factorization. What is done in the reference? Unfortunately, I cannot access it.

**Comment by abhinav28071999** [comment]
[Richard Crandall, Carl B. Pomerance - Prime numbers_ a computational perspective (2005, Springer).pdf](https://github.com/sympy/sympy/files/4716319/Richard.Crandall.Carl.B.Pomerance.-.Prime.numbers_.a.computational.perspective.2005.Springer.pdf)
The pseudocode is on pg344.

File: sympy/ntheory/ecm.py
Comment (abhinav28071999) [reply sympy/ntheory/ecm.py:18]: Shall i make them as `@staticmethod`

Comment (abhinav28071999) [reply sympy/ntheory/ecm.py]: Done.

File: sympy/ntheory/ecm.py
Comment (abhinav28071999) [reply sympy/ntheory/ecm.py]: we may also define `__mul__` method then we would not have to use `double` and `mont_ladder` outside the class.

**Comment by jksuom** [comment]
The `g == n` case for Pollard p-1 is discussed on p. 237. It says that "In the event of this type of failure, it is evident that increasing the search bound will not be of any help." Instead, a new start is necessary. It seems that that is needed also for ECM.

**Comment by abhinav28071999** [comment]
> The `g == n` case for Pollard p-1 is discussed on p. 237. It says that "In the event of this type of failure, it is evident that increasing the search bound will not be of any help." Instead, a new start is necessary. It seems that that is needed also for ECM.

Yup, i think this is needed for ecm.        if `g == n` then continue and try next curve.

**Comment by abhinav28071999** [comment]
@jksuom @jmig5776 any updates?

File: sympy/ntheory/ecm.py
Comment (jksuom) [inline sympy/ntheory/ecm.py:92]: I would have defined `add` and `double` as instance methods. Then this line could be `R = Q.double(n, a24)`. (It is not necessary to change the name of the first parameter in `double` to `self`, it can be `P`.)

Comment (abhinav28071999) [reply sympy/ntheory/ecm.py]: Done. Please take a look.

**Comment by abhinav28071999** [comment]
I have gone through `seed` but i am not sure how to use it here. My doubt is if we use seed for testing then how we will differ this seed when it is used outside of testing.

File: sympy/ntheory/ecm.py
Comment (Yathartha22) [inline sympy/ntheory/ecm.py:103]: two blank lines above.

**Review by Yathartha22 — CHANGES_REQUESTED** [review]
The PR seems fine. There is some work needed in terms of documentation and code quality. 

- Follow the blank line convention of pep8, you can refer [here](https://www.python.org/dev/peps/pep-0008/#blank-lines).
- Keep in mind the DRY (Do not Repeat Yourself) principle while writing the code.
- Code quality can be improved a bit further (use of intuitive variable names, more pythonic way of writing code etc.)

I will look at the implementation part in more detail.

**Comment by Yathartha22** [comment]
> I have gone through seed but i am not sure how to use it here ...

I think you can add `random.seed(1)` above [this](https://github.com/sympy/sympy/pull/19436/files#diff-3e19d4e345d3c5a0ca794de773fd685dR150) line, which would generate the same value from `6 to n` range.

**Comment by jksuom** [comment]
I think that the random seed should be set before the while loop. Then the sequence of values generated in the loop will be pseudorandom but stable to facilitate debugging.

File: sympy/ntheory/ecm.py
Comment (abhinav28071999) [reply sympy/ntheory/ecm.py:43]: I think space is not needed between `*` and `**`. Or am i missing something??

**Comment by abhinav28071999** [comment]
for seed i have  made it similar to pollard_rho, please check if it is correct.

**Comment by abhinav28071999** [comment]
please tell which documentation needs changes.

File: sympy/ntheory/ecm.py
Comment (abhinav28071999) [reply sympy/ntheory/ecm.py]: `(self.x - self.z)` is only being used one, so making another variable might be inefficient. I will check at other places where new variable can be introduced.

File: sympy/ntheory/ecm.py
Comment (abhinav28071999) [reply sympy/ntheory/ecm.py:77]: Q is a point whose starting value is self and if changes in the for loop.

**Comment by abhinav28071999** [comment]
any update?

File: sympy/ntheory/ecm.py
Comment (jmig5776) [inline sympy/ntheory/ecm.py:93]: `arthematic`->`arithematic`

**Comment by jmig5776** [comment]
I went through the research paper. According to that, the implementation seems to be fine but I would like you to more focus on the increase bound because I think it should be tweaked. Can you do more testing locally for getting optimal increase bound? Also proofread the documentation again so that we should not have grammatical errors at least.

File: sympy/ntheory/ecm.py
Comment (oscarbenjamin) [reply sympy/ntheory/ecm.py]: Neither of those spellings is correct :)

Comment (abhinav28071999) [reply sympy/ntheory/ecm.py]: corrected.

File: sympy/ntheory/ecm.py
Comment (jmig5776) [reply sympy/ntheory/ecm.py]: Space is needed to maintain homogeneity for operators throughout the code.

File: sympy/ntheory/ecm.py
Comment (jmig5776) [inline sympy/ntheory/ecm.py:222]: Suggested changes:
You can rename the `factor` to `_factors` and correspondingly rename `final` to `factors`.
And in the for loop I can be changed as ->` for i in sieve.primerange(1, 100000):`-> `for prime in sieve.primerange(1, 100000):`

File: sympy/ntheory/ecm.py
Comment (jmig5776) [inline sympy/ntheory/ecm.py:225]: I m bit confused. why are you checking divisibility of 2 as I think it will already be checked in the above for loop.

File: sympy/ntheory/ecm.py
Comment (jmig5776) [inline sympy/ntheory/ecm.py:236]: if you follow the above suggested change then you can rename `a` to `factor`.

File: sympy/ntheory/ecm.py
Comment (jmig5776) [inline sympy/ntheory/ecm.py:239]: This will also be changed to `for factor in _factors:`

**Comment by abhinav28071999** [comment]
@jksuom My next step is to improve documentation and add test case for Point class. Other than that are there any other changes required in this pr.

**Comment by jksuom** [comment]
I have not checked in detail with the reference but this looks plausible to me. The treatment of the random seed could be different though. Instead of setting it anew for each factor, it would probably suffice to set it once at the beginning in `ecm`.

**Comment by abhinav28071999** [comment]
I have improved the documentation.  Please review it.

**Comment by abhinav28071999** [comment]
> I went through the research paper. According to that, the implementation seems to be fine but I would like you to more focus on the increase bound because I think it should be tweaked. Can you do more testing locally for getting optimal increase bound? Also proofread the documentation again so that we should not have grammatical errors at least.

I have also my doubts here. Probably increase bound can be removed because it creates uncertainity of when the factors  will be found. One thing we can do is not to increase the bound indefinately and do it only once by a large factor maybe 10 or 100, then if it again fails the stop the operation.

File: sympy/ntheory/ecm.py
Comment (Yathartha22) [inline sympy/ntheory/ecm.py:17]: operation -> operations

File: sympy/ntheory/ecm.py
Comment (Yathartha22) [inline sympy/ntheory/ecm.py:41]: Add examples for this as well as other methods

**Review by Yathartha22 — COMMENTED** [review]
Some general comments
- Try to proofread the documentation so as to avoid unnecessarily typos.
- I see some comments are left unaddressed try to address them.
- also, try not to commit for every little change. Group similar changes into a single commit to avoid unnecessary commits in the log.

Also, could you share a snap of the performance comparison of `ecm` and `factorint`

**Comment by abhinav28071999** [comment]
image
For the last one i waited for 10mins but factorint  was not able to factorize.

**Comment by abhinav28071999** [comment]
image

**Comment by abhinav28071999** [comment]
I am getting some problem in creating test cases for Point class. Initially i thought to use elliptic_curve function but the problem is that the elliptic_curve function the z_coordinate is always 1 but in this Point class the z_cord is not necessarily 1. I will do some research.

**Comment by abhinav28071999** [comment]
the main problem that i am getting          is that nowwhere i can find addition in 3 coordinates in elliptic curve.

**Comment by abhinav28071999** [comment]
i guess i will just use this formula to make test cases.
image

**Comment by abhinav28071999** [comment]
@Yathartha22 @jmig5776 i have added test case for double   method. If  this way seems to be correct then i will add for add method also.

**Comment by abhinav28071999** [comment]
i am stuck in a problem. The double method seems good, but in addition method the equation in the book and the paper seems different.  I expanded the formula in the paper but it has additional multiplier  `4`. I was wondering which one i should use.
image

**Comment by jksuom** [comment]
This algorithm is intended for numbers `n` that have no small divisors. In particular, `n` is odd, and therefore the factor 4 is invertible mod `n`. Both methods lead to the same sequence of points, only their homogeneous coordinates differ by a factor invertible mod `n`.

It seems that the method of Gaj.pdf is designed under the assumption that multiplication is significantly more expensive than addition or subtraction. There are two products that are used on both line 1 and line 2. The other algorithm need several multiplications for the same coordinates but the formulas are simpler. I would try to find the time taken by these operations and base the choice on the results.

**Comment by abhinav28071999** [comment]
By using method from Gaj i am getting improvement of around 20%.

**Comment by abhinav28071999** [comment]
one thing that i am unsure is that should the Point class be made public or not.

**Comment by abhinav28071999** [comment]
> I have not checked in detail with the reference but this looks plausible to me. The treatment of the random seed could be different though. Instead of setting it anew for each factor, it would probably suffice to set it once at the beginning in `ecm`.

I was thinking of using `ecm_one_factor` in factorint, so then the random seed should also be set in this function.

**Comment by abhinav28071999** [comment]
@Yathartha22 I have addressed all the comments. If there is any left please let me know. Also before merging i will squash the commits.

**Comment by Yathartha22** [comment]
Great! I will have a look at it. @jksuom what are your thoughts on this. Would you mind having a detailed review of the implementation part? Are there any shortcomings/leftovers that can be taken care of here/in future PR's? We would need your final approval before this gets in.

File: sympy/ntheory/__init__.py
Comment (jksuom) [inline sympy/ntheory/__init__.py:28]: `Point` is a fairly special object: no `y` coordinate, includes `a_24` that is probably meaningless outside of this module. I would prefer not making `Point` public.

Comment (abhinav28071999) [reply sympy/ntheory/__init__.py:28]: I was also thinking this. One mere thing shall i remove the examples and test cases of Point class entirely.

**Comment by jksuom** [comment]
I did not check carefully with the reference but I trust that the implementation is correct as it seems to work. There remains the task of adding `ecm` to `factorint`. It is more expensive than the elementary methods so it should be called only after those fail. The bounds may have to be adjusted according to the size of the integer to be factored. It should probably be also checked that it is not a prime.

**Comment by abhinav28071999** [comment]
@jksuom One more thing that i am confused is that if  so we really need `increase_bound` parameter, if we need that then i think the multiplier can be tweaked and probably the increase should not be done indefinately, maybe once and twice and if we don't factorize then just give up.

**Comment by jksuom** [comment]
The algorithm for one factor can fail in two different ways: either `g` is `1` or `g` is `n`. The algorithm is based on the idea that even though the prime factors `p` of `n` can be very big, the number of points of some elliptic curve `E_p` associated with `p` may have all its prime factors smaller than some bound `B2`. Then order of the group `E_p` is a factor of some big number `k` that is a product of primes below `B2` and the multiple `k*P` of any point will be the neutral element of `E_p (with `z` coordinate 0 mod `p`).

Now, if the one factor algorithm fails with `g == 1`, then the orders of all groups `E_p` have prime factors bigger than `B2`. This can happen if `n` is a big prime. But, if it is not a prime, then it is reasonable to repeat the algorithm with another elliptic curve. It should not be necessary increase `B2` if it has been reasonably chosen in the beginning. Several curves can be tested.

If the one factor algorithm fails with `g == n`, then the multiple `k*P` is the neutral element for all prime factors of `n`. This means that `k` is too big and should be reduced instead of increasing the bound `B2`. In practice, it is probably best to choose another curve again.

So it seems that, in either case, `increase_bound` is not needed.

**Comment by abhinav28071999** [comment]
@jksuom  @Yathartha22  since `Point` class is not going to be made public, shall i remove the examples and test cases of the point class.

**Comment by Yathartha22** [comment]
> shall i remove the examples and test cases of the point class.

Regarding the examples, I think there's no harm keeping the examples as it is, it would increase the understanding of the class.
And for the tests, they should be added regardless of anything being public/private.

File: sympy/ntheory/__init__.py
Comment (abhinav28071999) [inline sympy/ntheory/__init__.py:28]: For making this private, i should remove `Point`, but how will i import `Point`  for test_ecm.py.

Comment (jksuom) [reply sympy/ntheory/__init__.py:28]: It can be imported directly from ecm: `from sympy.ntheory.ecm import ecm, Point`

**Comment by abhinav28071999** [comment]
All the checks have passed. Before merging  let me know, i can squash some commits.

File: sympy/ntheory/ecm.py
Comment (Yathartha22) [inline sympy/ntheory/ecm.py:52]: I don't get this, is `diff` really the difference between two points, or some random point (as shown in the example) if so how it is defined and why can't it be calculated within the method itself? and if not then the name is ambiguous.

Also using `P` is not obvious, we can replace it with `self`.

Comment (abhinav28071999) [reply sympy/ntheory/ecm.py:52]: yes,   diff is the difference between the 2 points

File: sympy/ntheory/ecm.py
Comment (jksuom) [reply sympy/ntheory/ecm.py:263]: It could have default value `None` in which case the code would not set the seed. But it could be useful to have in case someone would want to have a repeatable procedure for optimizing the parameters, for example. This function is not intended for ordinary users. Only `factorint` would invoke `ecm` under certain conditions.

File: sympy/ntheory/ecm.py
Comment (abhinav28071999) [reply sympy/ntheory/ecm.py:294]: u have previously one this, but https://github.com/sympy/sympy/pull/19436#issuecomment-641394666 suggested this way, which one is more appropriate??

Comment (abhinav28071999) [reply sympy/ntheory/ecm.py:294]: And in the example i am not using diff as a random point. I have taken  a  initial point p1, then p2 is actually 2*p1. Then p3 = p2 + p1, here the diff = p2-p1 = 2*p1 - p1 = p1,
so p2.add(p1, diff=p1) gives p3.

File: sympy/ntheory/ecm.py
Comment (abhinav28071999) [reply sympy/ntheory/ecm.py:52]: And in the example i am not using diff as a random point. I have taken a initial point p1, then p2 is actually 2p1. Then p3 = p2 + p1, here the diff = p2-p1 = 2p1 - p1 = p1,
so p2.add(p1, diff=p1) gives p3.

Comment (Yathartha22) [reply sympy/ntheory/ecm.py:52]: In that case, we can calculate it in the method itself, isn't it?

Comment (abhinav28071999) [reply sympy/ntheory/ecm.py]: No, we cannot. the  useful of `add` is that we can add 2 points   using less multiplications if we already know the difference between them.

Comment (abhinav28071999) [reply sympy/ntheory/ecm.py]: in the example i made the point such that it is `2*p1`. But  in general we don't know about this. You can observe that whereever i use `add` we knew what the difference of the points was. For example, in the `mont_ladder` it is constructed such a way that the difference  between any twp intermediate points was the initial point.

**Comment by abhinav28071999** [comment]
@jksuom @Yathartha22 Take a look here. It tells that lucas chains can be used instead of montgomery ladder algorithm.
image

**Comment by abhinav28071999** [comment]
any updates?

File: sympy/ntheory/ecm.py
Comment (Yathartha22) [reply sympy/ntheory/ecm.py:293]: Okay! I see it makes more sense here.

File: sympy/ntheory/ecm.py
Comment (Yathartha22) [reply sympy/ntheory/ecm.py]: Alright! then perhaps we need to mention this in the docstring as well that why `diff` is not computed inside the method.

**Comment by jksuom** [comment]
It looks like multiples in a Lucas sequence can be computed more efficiently than using the binary expansion but can that method be used for multiples `k*P`` where `k` should be divisible by all primes below some bound?

File: sympy/ntheory/ecm.py
Comment (abhinav28071999) [reply sympy/ntheory/ecm.py]: I will do that.

**Comment by abhinav28071999** [comment]
> It looks like multiples in a Lucas sequence can be computed more efficiently than using the binary expansion but can that method be used for multiples `` k*P`` where ``k` should be divisible by all primes below some bound?

I have also not figured it out. The thing is for fibonacci seq to work `k` should lie to be in tat, but its not always true.

File: sympy/ntheory/ecm.py
Comment (abhinav28071999) [reply sympy/ntheory/ecm.py]: I am having doubt, the diff is already known to us, so it seems conflicting      to mention that diff cannot be calculated inside the function.

Comment (Yathartha22) [reply sympy/ntheory/ecm.py]: >  it seems conflicting to mention that diff cannot be calculated inside the function.

I meant to reason out "why" we are not calculating `diff` within the method (which is not obvious in general) --> so as to limit the number of multiplication operations, as you mentioned above.

Comment (Yathartha22) [reply sympy/ntheory/ecm.py]: > the useful of add is that we can add 2 points using less multiplications if we already know the difference between them.

A more convincing reason for this is mentioned int the`Gaj.pdf`, below Algorithm 3, states

> The algorithm is constructed in such a way that the difference between the intermediate points P and Q, P −Q,
is always constant, and equal to the value of the initial point P0.

This makes things more clear on taking `diff` as a parameter because we would always know the `diff`. This I guess needs to be mentioned in the docstring.

File: sympy/ntheory/ecm.py
Comment (Yathartha22) [inline sympy/ntheory/ecm.py:14]: I think we can mention the general form of the equation here.

**Comment by Yathartha22** [comment]
> I have also not figured it out. The thing is for fibonacci seq to work k should lie to be in tat, but its not always true.

Did you conclude on this? Otherwise, we should stick to the current more general algorithm?

**Comment by abhinav28071999** [comment]
> > I have also not figured it out. The thing is for fibonacci seq to work k should lie to be in tat, but its not always true.
> 
> Did you conclude on this? Otherwise, we should stick to the current more general algorithm?

For now i think we should stick to the general algorithm. I was researching about lucas sequence but for this case i couldn't find any. My guess is that if we use different starting point in fibonacci series such that we reach `k` but this will be difficult to find the starting points. Anyway, such optimizations can be done in a future pr.

File: sympy/ntheory/ecm.py
Comment (Yathartha22) [inline sympy/ntheory/ecm.py:63]: typo algorithm, `this` redundancy.
Also, look to my latest comment for this reason.

**Comment by Yathartha22** [comment]
I have cancelled the build to avoid unnecessary delay for other Pr's, you would need another commit to fix the recent changes

**Comment by Yathartha22** [comment]
I guess the PR seems to be in good shape now.
+1.

**Comment by abhinav28071999** [comment]
@jksuom @Yathartha22 is this ready to be merged?

**Comment by Yathartha22** [comment]
I will merge this in 24 hrs if no one has no issues.
@jksuom

**Review by Yathartha22 — APPROVED** [review]
Great! 
@abhinav28071999 thanks for your contribution. I will be merging this now.
I guess there is one thing that needs to be done: to use `ecm` instead of `factorint` as discussed. You can make another PR for this.

**Pull request merged**
